src/nft.sol

```diff
@@ -32,166 +32,163 @@ contract IdentityFactoryLike {
+     * @param values bytes[] value associated with each property
+     * @param salts bytes[] salt associated with each property
+     */
+    function _verifyAsset(address to, bytes[] memory properties, bytes[] memory values, bytes32[] memory salts) internal view  {
+        // construct assetHash from the props, values, salts
+        // append to address
+        bytes memory data = abi.encodePacked(to);
+
+        // append hashes
+        for (uint i=0; i< properties.length; i++) {
+            data = abi.encodePacked(data, keccak256(abi.encodePacked(properties[i], values[i], salts[i])));
         }
-        return false;
+
+        assetManager.verify(keccak256(data));
+    }
+
+    /**
+     * @dev Checks that provided document is signed by the given identity. Validates and checks if the public key used is a not revoked and has the purpose SIGNING_KEY.
+     * Does not check if the signature root is part of the document root.
+     * @param identity address identity which is a collaborator of the document
+     * @param dataRoot bytes32 hash of all data fields of the document which are signed
+     * @param signature bytes contains signature + transition flag
+     */
+    function _signed_document(address identity, bytes32 dataRoot, bytes memory signature) internal view {
+        // check that the identity being used has been created by the Centrifuge Identity Factory contract
+        require(identityFactory.createdIdentity(identity), "nft/identity-not-registered");
+        // extract flag from signature
+        (bytes memory sig, byte flag) = _recoverSignatureAndFlag(signature);
+        // Recalculate hash and extract the public key from the signature
+        address key = _toEthSignedMessage(dataRoot, flag).recover(sig);
+        bytes32 pubKey = bytes32(uint(key) << 96);
+        // check that public key has signature purpose on provided identity
+        KeyManagerLike keyManager = KeyManagerLike(identity);
+        require(keyManager.keyHasPurpose(pubKey, SIGNING_PURPOSE), "nft/sig-key-not-valid");
+        (, , uint32 revokedAt) = keyManager.getKey(pubKey);
+        require(revokedAt == 0, "nft/key-revoked");
     }
 
-    // --- NFT ---
-    function _checkAnchor(uint anchor, bytes32 data_root, bytes32 sig_root) internal view returns (bool) {
-        bytes32 doc_root;
-        (, doc_root, ) = anchors.getAnchorById(anchor);
-        if (data_root < sig_root) {
-           return doc_root == sha256(concat(data_root, sig_root));
-        } else {
-           return doc_root == sha256(concat(sig_root, data_root));
+    function _recoverSignatureAndFlag(bytes memory signature) internal pure returns (bytes memory, byte) {
+        // ensure signature value is 66
+        require(signature.length == 66, "nft/invalid-signature-length");
+        byte flag = signature[65];
+        bytes memory sig = new bytes(65);
+        uint i = 0;
+        for (i = 0; i < 65; i++) {
```

Comment:
can you optimize this?
https://stackoverflow.com/questions/49051856/is-there-a-pop-functionality-for-solidity-arrays

try:
```javascript
delete myArray[myArray.length - 1];
myArray.length--;
```

Reply:
I have tried this before. This would not work since memory array cannot be resized. And another option is to have the array as storage but that would cost gas. Hence, this for loop @manuelpolzhofer

Reply:
We can go with assembly route, I'm not very proficient on that, but still, doubt its worthiness regarding the optimisations